src/utils.jl

```diff
@@ -50,6 +50,31 @@ function centercols!(m::AbstractMatrix)
     return m
 end
 
+# generic function for calculating the
+# criterion and the gradient of the method
+# based on the weighted sums of columns and rows
+function weighted_sums_criterion_and_gradient!(
```

Comment:
I think `weighted_sums_criterion_and_gradient!` should live in src/methods/methods.jl instead.